reconcile.go

```diff
@@ -232,56 +260,70 @@ func (a *Autopilot) pruneDeadServers() error {
+	state := a.GetState()
 
-	// remove failed non voting servers
-	for _, srv := range failed.FailedNonVoters {
-		a.logger.Info("Attempting removal of failed server node", "id", srv.ID, "name", srv.Name, "address", srv.Address)
-		a.delegate.RemoveFailedServer(srv)
+	// Support not breaking the promoter's interface for filtering servers
+	failedServers := servers.convertToFailedServers(state)
+	failedServers = a.promoter.FilterFailedServerRemovals(conf, state, failedServers)
+	servers = servers.convertFromFailedServers(failedServers)
+
+	// Partially apply (wrap) some functions for clarity below
+	removeStale := removeStale(a)
+	removeFailed := removeFailed(a)
+	adjudicate := func(logger hclog.Logger, minQuorum uint, voterCountProvider func() int) func(RaftServers) []raft.ServerID {
+		return func(raftServers RaftServers) []raft.ServerID {
+			return adjudicateRemoval(logger, minQuorum, voterCountProvider, raftServers)
+		}
+	}(a.logger, conf.MinQuorum, servers.PotentialVoters)
+
+	// Try to remove servers in order of increasing precedence
+
+	// Remove all stale non-voters
+	if err = remove(removeStale, adjudicate(servers.StaleNonVoters)); err != nil {
+		return err
 	}
 
-	// remove stale non voters
-	for _, id := range failed.StaleNonVoters {
-		a.logger.Debug("removing stale raft server from configuration", "id", id)
-		if err := a.removeServer(id); err != nil {
-			return err
-		}
+	// Remove stale voters
+	if err = remove(removeStale, adjudicate(servers.StaleVoters)); err != nil {
+		return err
 	}
 
-	maxRemoval := (voters - 1) / 2
+	// Remove failed non-voters
+	if err = remove(removeFailed, adjudicate(servers.FailedNonVoters)); err != nil {
+		return err
+	}
 
-	for _, id := range failed.StaleVoters {
-		if voters-1 < int(conf.MinQuorum) {
-			a.logger.Debug("will not remove server as it would leave less voters than the minimum number allowed", "id", id, "min", conf.MinQuorum)
-		} else if maxRemoval < 1 {
-			a.logger.Debug("will not remove server as removal of a majority of servers is not safe", "id", id)
-		} else if err := a.removeServer(id); err != nil {
-			return err
-		} else {
-			maxRemoval--
-			voters--
-		}
+	// Remove failed voters
+	if err = remove(removeFailed, adjudicate(servers.FailedVoters)); err != nil {
+		return err
 	}
 
-	for _, srv := range failed.FailedVoters {
-		if voters-1 < int(conf.MinQuorum) {
-			a.logger.Debug("will not remove server as it would leave less voters than the minimum number allowed", "id", srv.ID, "min", conf.MinQuorum)
-		} else if maxRemoval < 1 {
-			a.logger.Debug("will not remove server as a removal of a majority of servers is not safe", "id", srv.ID)
+	return nil
+}
+
+func adjudicateRemoval(logger hclog.Logger, minQuorum uint, voterCountProvider func() int, s RaftServers) []raft.ServerID {
+	var ids []raft.ServerID
+	failureTolerance := getFailureTolerance(voterCountProvider())
+
+	for id, v := range s {
+		if failureTolerance < 1 {
```

Comment:
If failure tolerance is less than 1, removing a read replica would still be fine. So the logic seems a little off here.